tests/runner.py

```diff
@@ -275,7 +280,37 @@ def create_test_file(name, contents, binary=False):
 test_index = 0
 
 
-class RunnerCore(unittest.TestCase):
+def parameterized(parameters):
+  def decorator(func):
+    func._parameterize = parameters
+    return func
+  return decorator
+
+
+class RunnerMeta(type):
+  @classmethod
+  def make_test(mcs, name, func, suffix, args, kwargs={}):
+    @wraps(func)
+    def resulting_test(self):
+      return func(self, *args, **kwargs)
+    resulting_test.__name__ = '%s_%s' % (name, suffix)
+    if hasattr(func, '__qualname__'):
+      resulting_test.__qualname__ = '%s_%s' % (func.__qualname__, name)
+    return resulting_test.__name__, resulting_test
+
+  def __new__(mcs, name, bases, attrs):
+    new_attrs = {}
+    for attr_name, value in attrs.items():
+      if hasattr(value, '_parameterize'):
+        for parameter in value._parameterize:
+          new_name, func = mcs.make_test(attr_name, value, *parameter)
+          new_attrs[new_name] = func
+      else:
+        new_attrs[attr_name] = value
+    return type.__new__(mcs, name, bases, new_attrs)
```

Comment:
So, I really don't have a good understanding of metaclasses, so I don't know how appropriate it is in this case. The concern is, how hard is this code to change? How much unrelated python knowledge does someone need to be able to tweak the metadce tests?

In principle I'm a fan of using all the language features, especially if their complexity can be abstracted away. I'm not usually on the side of the table where I don't understand the feature, so this is educational 😀. In particular, it's hard to reason about how well the complexity is hidden here, with this partial understanding.

It looks like a clean interface from a test reader's perspective, in that we just have an annotation that magically works. Trying to understand how the magic works leads to lots of hidden complexity, so I'm not sure how well this is abstracted.